### PR TITLE
Split Celery logs into stdout/stderr

### PR DESCRIPTION
**Pull request by apilaskowski**
Celery logs everything into Error stream, which is probably sufficient for most use-cases.
It might be more intuitive and more desired to have errors/criticals go to stderr, while low severity informations like info/warning go to stdout.

I have tested this solution, which I propose here with worker producing various severity levels of logs and they are correctly split into out and error streams.

I provided a solution which is backward compatible, so anyone that does not want to use it can just ignore it. 

During pre-commit I discovered that config.yml needs to be updated and I am not sure how to perform that correctly.
We need to adjust config.yml for this to be complete, I wasn't sure if this should be made by hand or via some script.

### CODE REVIEW

**Comment by apilaskowski** [comment]
Can I somehow request Reviewers, or are reviewers assigned automatically? 
@kosteev @bjankie1

**Comment by apilaskowski** [comment]
@potiuk Do you mean to update `airflow/config_templates/config.yml` and then pre-commit will auto-generate `default-airflow.cfg'? 
Should I set version to 2.6.0?
I think that now it is ok (I updated `config.yml` and `default-airflow.cfg` was automatic).

File: airflow/cli/commands/celery_command.py
Comment (kosteev) [inline airflow/cli/commands/celery_command.py]: I think you can do it in one call like this which should be the same
```
if conf.get("logging", "reset_and_split_logging", fallback=None):
```

File: airflow/cli/commands/celery_command.py
Comment (kosteev) [inline airflow/cli/commands/celery_command.py]: Which formatter is actually used by Celery at the moment?
I think they have some different formatter by default, but if we are making this change probably it is good idea to format Celery logs with same formatter.

then you can do settings.LOG_FORMAT here?

File: airflow/cli/commands/celery_command.py
Comment (kosteev) [inline airflow/cli/commands/celery_command.py]: Another way would be to do
```
logger.handlers[:] = [below_error_handler, from_error_handler]
```
(as done in _capture_task_logs method)

Since we access `logger.handlers` in `logger.handlers.clear()` anyway, then above can be more concise version of achieving the same.

File: airflow/config_templates/config.yml
Comment (kosteev) [inline airflow/config_templates/config.yml]: This should somehow indicate that it is about Celery logs.

File: airflow/config_templates/config.yml
Comment (kosteev) [inline airflow/config_templates/config.yml]: IMHO this doesn't sound very explicit that if it is enabled then logs with WARNING level and below are sent to stdout .... "It might be useful" a bit confusing for me

**Review by kosteev — APPROVED** [review]

**Comment by potiuk** [comment]
> @potiuk Do you mean to update `airflow/config_templates/config.yml` and then pre-commit will auto-generate `default-airflow.cfg`?

Ah yeah. I copied  **wrong** file. But you figured it out it seems :)

File: airflow/cli/commands/celery_command.py
Comment (uranusjr) [inline airflow/cli/commands/celery_command.py]: Why do these need to be aliased? Doesn’t `import sys` and `sys.stdout` `sys.stderr` do the job?

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py]: I think it is this one:
```"""[%(asctime)s: %(levelname)s/%(processName)s] %(task_name)s[%(task_id)s]: %(message)s"""```
From:
https://github.com/celery/celery/blob/025c1d4f9a1208a9d3bb01b1129a0f48234d3df0/celery/app/defaults.py#L17

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py]: I intended it to be easier distinguishable from `args.stderr` etc.

Inside the `worker` function there is a local variable, which is called `stderr` and when I was working on that I wanted to have them easy to differentiate. 

If it is ok, I will use version with `sys.stderr` instead.

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py]: Updated.

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py]: Updated.

File: airflow/config_templates/config.yml
Comment (apilaskowski) [reply airflow/config_templates/config.yml]: What do you think about:
`celery_logging_override`

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py]: Updated.

File: airflow/config_templates/config.yml
Comment (apilaskowski) [reply airflow/config_templates/config.yml]: Do you think that:

```
By default Celery sends all logs into stderr.
If this option is selected AirFlow will handle and send low level logs like INFO and WARNING to stdout,
while sending higher severity logs to stderr.
This option removes default logging handlers, and provide new ones.
```

Explains it better, or should I give some more information there?

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py]: Should I put it directly into code, or should it be in the configuration file?

File: airflow/config_templates/config.yml
Comment (kosteev) [reply airflow/config_templates/config.yml]: I see "option is enabled" used instead of "option is selected" in other places

Also I would change "This option removes default logging handlers, and provide new ones." to
"If enabled default logging handlers will be removed, and new one provided."

File: airflow/config_templates/config.yml
Comment (kosteev) [reply airflow/config_templates/config.yml]: I like it as it is consistent with "celery_logging_level".

What do you think about "celery_logging_split" to have more indication of the way how it is overriden?
@potiuk what are your thoughts?

File: airflow/cli/commands/celery_command.py
Comment (kosteev) [reply airflow/cli/commands/celery_command.py]: I would use either DEFAULT_TASK_LOG_FMT const or settings.LOG_FORMAT const here

File: airflow/config_templates/config.yml
Comment (apilaskowski) [reply airflow/config_templates/config.yml]: Updated.

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py]: Please let me know whether current solution is ok.

File: airflow/config_templates/config.yml
Comment (apilaskowski) [reply airflow/config_templates/config.yml]: I updated it to *_split. 

Let's wait for additional confirmation before resolving this topic.

File: airflow/cli/commands/celery_command.py
Comment (kosteev) [inline airflow/cli/commands/celery_command.py]: why do we copy it? why can't we import it
```
from celery.app.defaults import DEFAULT_TASK_LOG_FMT
```

File: airflow/cli/commands/celery_command.py
Comment (kosteev) [inline airflow/cli/commands/celery_command.py]: I believe accord to pep8 it should be just "import sys".
In the code you still should use it as sys.stderr/sys.stdout

Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py]: I was always trying to keep imports as narrow as possible, but I am ok with that change.
Updated.

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py]: We can. I did now.
Updated.

File: airflow/config_templates/config.yml
Comment (apilaskowski) [reply airflow/config_templates/config.yml]: @potiuk @uranusjr What do you think, is this name good?

**Comment by apilaskowski** [comment]
@potiuk @uranusjr @kosteev What do you think? Can we merge this PR?

**Comment by apilaskowski** [comment]
I updated some small problems following pre-commit recomendations.

File: airflow/cli/commands/celery_command.py
Comment (uranusjr) [inline airflow/cli/commands/celery_command.py]: This should use a `getboolean` instead

File: airflow/cli/commands/celery_command.py
Comment (uranusjr) [inline airflow/cli/commands/celery_command.py:120]: How is this different from `setLevel`?

File: airflow/cli/commands/celery_command.py
Comment (uranusjr) [inline airflow/cli/commands/celery_command.py:127]: I wonder if it’s a good idea to replace the handlers entirely. Is it possible to have the logger already pointing to somewhere else e.g. a file?

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py:120]: I intend for only low level logging to go to stdout.
As far as I understand it `setLevel` sets threshold so lower levels would be ignored, while higher not.
I would like to have levels of error and above to be ignored by this handler.

Is there better way to send only low level logging to stdout? 
This is the solution that I come up with, but there might be some better one.

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py]: Updated.

File: airflow/cli/commands/celery_command.py
Comment (uranusjr) [reply airflow/cli/commands/celery_command.py:120]: I see, I misunderstood. Perhaps a docstring in `NoErrorOrAboveFilter` could help prevent the same error by others?

Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py:120]: Great point. I will suggest one ASAP.

Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py:120]: @uranusjr what do you think now?

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py:127]: I was also thinking about it.
I don't think following solution is very clean, but if we want we could do:

- go through all previous handlers
- remove any StreamHandler, because we do not want duplicate logs
- leave any FileHandlers, if they are present then user probably introduced them purposefully 

Alternatively, if you think it is better I would suggest to put more weight on the fact that those handlers are cleared in documentation and docstring, so user selecting this flag will expect that only those two handlers will be present.

File: airflow/cli/commands/celery_command.py
Comment (uranusjr) [inline airflow/cli/commands/celery_command.py]:
```suggestion
            """Allow only logs *lower* than warning to be reported."""
```

File: airflow/cli/commands/celery_command.py
Comment (uranusjr) [reply airflow/cli/commands/celery_command.py:120]: I flipped the wording to a positive tone (I think it avoids some mental overhead), and also switch to `< logging.ERROR` instead. Using `<= logging.WARNING` is problematic; it is possible to emit a log with an in-between level, e.g. `log(35, "...")` that would be missed.

File: airflow/cli/commands/celery_command.py
Comment (uranusjr) [inline airflow/cli/commands/celery_command.py]:
```suggestion
            """Allow only logs with level warning or lower to be reported."""
```

File: airflow/cli/commands/celery_command.py
Comment (uranusjr) [reply airflow/cli/commands/celery_command.py:127]: Hmm yeah the workaround doesn’t seem very reliable either. Maybe we can rename the config to something more specific (that describes the exact “error to stderr, lower to file” behaviour) and describe it in the documentation.

File: airflow/cli/commands/celery_command.py
Comment (uranusjr) [inline airflow/cli/commands/celery_command.py]:
```suggestion
        class NoErrorOrAboveFilter(logging.Filter):
            """Allow only logs with level *lower* than ERROR to be reported."""

            def filter(self, record):
                return record.levelno < logging.ERROR
```

File: airflow/cli/commands/celery_command.py
Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py:127]: Maybe`celery_logging_reset_and_split`?
I also updated docstring and config.yml so it highlights a bit more the fact that those previous handlers will get removed. Do you think that it is sufficient?

Comment (uranusjr) [reply airflow/cli/commands/celery_command.py:127]: Not sure about the config, but yeah the rest looks good to me.

Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py:127]: Do you have an intuition what might be missing, or badly explained?
Should we have another opinion on config topic?
Let me know if I can try to fix it.

Comment (uranusjr) [reply airflow/cli/commands/celery_command.py:127]: Mostly “reset and split” seems vague to me—what does this reset logging to and split it into? I don’t have a concrete suggestion but feel like there should be a better name.

Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py:127]: Another propositions:

`celery_logging_remove_default_handlers_and_use_custom`
`celery_logging_use_custom_handlers`

However, forst one is quite long.

Comment (kosteev) [reply airflow/cli/commands/celery_command.py:127]: `logging` in the name of property looks redundant for me as we have `logging` as a section.
Idea from me: `celery_logs_in_stdout_stderr`, we still have `logs` in the name but omitting can be confusing.

Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py:127]: @uranusjr 
What do you think?

Comment (potiuk) [reply airflow/cli/commands/celery_command.py:127]: Maybe just `celery_stdout_stderr_separation` ?

**Review by potiuk — APPROVED** [review]
LGTM. Pending working out the hardest part - i.e. naming :D

File: airflow/cli/commands/celery_command.py
Comment (potiuk) [reply airflow/cli/commands/celery_command.py:127]: WDYT @uranusjr ?

Comment (apilaskowski) [reply airflow/cli/commands/celery_command.py:127]: IMHO `celery_stdout_stderr_separation` express the spirit of this option.
As far as I know (I contacted @uranusjr) he is OOO for at least a week more.

Comment (potiuk) [reply airflow/cli/commands/celery_command.py:127]: let's go ahead wth this then - we can always fix it later.

**Comment by potiuk** [comment]
First thing I do with every single PR of mine when I come back to it is to rebase it on top of the latest main and I heartily recommend making it into a habit to everyone:

This:

image

means that we absolutely should not merge that one because what runs in your CI pipeline now might fail badly when merged. Rebase early - rebase often is something I keep on repeating.

**Comment by potiuk** [comment]
No strict need to rebase (though I personally prefer it).  If you merged and it looks good after conflict solving, it will work as welll (we follow "Squash and merge" workflow when merging, which means that all PRs are squashed into a single commit when merging, so it does not really matter if you rebased or merged your PR as long as it worked and conflicts were resolved.

**Comment by apilaskowski** [comment]
Can you confirm that what pre-commit did is correct?
Because I am not sure if those files should be there or not.
They were added by pre-commit command. (_vendors dir).

**Comment by potiuk** [comment]
No - it is not good. You should remove them - they were only added because old (226 commits behind) pre-commit version have no exclusion for that folder.

**Comment by potiuk** [comment]
(BTW. This is why "rebase early/rebase often" is always a good strategy.

**Comment by apilaskowski** [comment]
@potiuk I think that now everything is working properly :)

**Comment by potiuk** [comment]
> @potiuk I think that now everything is working properly :)

And @uranusjr is around it seems for a quick check :)

**Comment by apilaskowski** [comment]
@uranusjr @potiuk 
Let me know if/when we are ready.
I will update the branch with a few new commits that occurred in a mean time (I don't want to use CI resources to much, or should I?).

**Pull request merged**
